src/ring_theory/etale.lean

```diff
@@ -0,0 +1,224 @@
+/-
+Copyright (c) 2022 Andrew Yang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Andrew Yang
+-/
+import ring_theory.ideal.operations
+import ring_theory.nilpotent
+import ring_theory.tensor_product
+import linear_algebra.isomorphisms
+
+/-!
+
+# Formally étale morphisms
+
+An `R` algebra `A` is formally étale (resp. unramified, smooth) if for every `R`-algebra,
+every square-zero ideal `I : ideal B` and `f : A →ₐ[R] B ⧸ I`, there exists
+exactly (resp. at most, at least) one lift `A →ₐ[R] B`.
+
+We show that the property extends onto nilpotent ideals, and that these properties are stable
+under `R`-algebra homomorphisms and compositions.
+
+-/
+
+universes u
+
+namespace algebra
+
+section
+
+variables (R : Type u) [comm_semiring R]
+variables (A : Type u) [semiring A] [algebra R A]
+variables {B : Type u} [comm_ring B] [algebra R B] (I : ideal B)
+
+include R A
+
+/-- An `R` algebra `A` is formally unramified if for every `R`-algebra, every square-zero ideal
+`I : ideal B` and `f : A →ₐ[R] B ⧸ I`, there exists at most one lift `A →ₐ[R] B`. -/
+def formally_unramified : Prop :=
```

Comment:
Is there a reason why these are not classes?

Reply:
Yes, but [algebra.finite_type](https://leanprover-community.github.io/mathlib_docs/ring_theory/finiteness.html#algebra.finite_type) is. Maybe you can do the same: as a class for algebras and a def for morphisms. It is what it is done in ring_theory.finiteness.

Reply:
This is done already. I forgot to resolve this.